core/app/models/spree/promotion.rb

```diff
@@ -193,6 +193,7 @@ def usage_count(excluded_orders: [])
       Spree::Adjustment.promotion.
         eligible.
         in_completed_orders(excluded_orders: excluded_orders).
+        where(spree_orders: { id: Spree::Order.not_canceled }).
```

Comment:
I think there may be performance issues associated with this subquery, as it yields all orders in the db except for the ones canceled. This could mean millions of records on large enough DBs.

Reply:
@spaghetticode, thanks for catching this 🙏 

I have tried an alternative approach that changes the `#in_completed_orders` method in `Adjustments`. 
It does not alter the method's default behaviour, but it allows for `canceled` orders to be excluded on passing an optional argument `exclude_canceled_orders`. 
Please see this [commit](https://github.com/solidusio/solidus/pull/4123/commits/1745930de91e1b14a80c0190ad2d06512bbb3789).

Although the old `#usage_count` behaviour can be considered a bug, we can also use a config flag with this change if you feel this is the better way to go.

It does look a bit messy to me, so please let me know if this is okay and if there is a better way to do something like conditional method changing in this case.

I see that I used the non-American spelling of `canceled`(`cancelled`) in the argument 🤦. I will fix that, squash the commits and put some more of this context into the commit message when you confirm this approach is fine.


The query that was running before you raised this issue was:
```sql
SELECT COUNT(DISTINCT "spree_adjustments"."order_id") FROM "spree_adjustments" INNER JOIN "spree_orders" ON "spree_orders"."id" = "spree_adjustments"."order_id" WHERE "spree_adjustments"."promotion_code_id" = ? AND "spree_adjustments"."eligible" = ? AND "spree_orders"."completed_at" IS NOT NULL AND 1=1 AND "spree_orders"."state" != ?  [["promotion_code_id", 1], ["eligible", 1], ["state", "canceled"]]
```

With the new approach, it runs this instead:
```sql
SELECT COUNT(DISTINCT "spree_adjustments"."order_id") FROM "spree_adjustments" INNER JOIN "spree_orders" ON "spree_orders"."id" = "spree_adjustments"."order_id" WHERE "spree_adjustments"."promotion_code_id" = ? AND "spree_adjustments"."eligible" = ? AND "spree_orders"."completed_at" IS NOT NULL AND 1=1 AND "spree_orders"."id" IN (SELECT "spree_orders"."id" FROM "spree_orders" WHERE "spree_orders"."state" != ?)  [["promotion_code_id", 1], ["eligible", 1], ["state", "canceled"]]
```

Reply:
Great! Thanks for double checking. I think I pasted the same query String in both places 🤦